### PR TITLE
nested task and subject form

### PR DESCRIPTION
**Pull request by Hoang-Tung**

### CODE REVIEW

File: app/controllers/supervisor/subjects_controller.rb [hunk 1]
```diff
@@ -12,6 +12,7 @@ def show
 
   def new
     @subject = Subject.new
+      2.times {@subject.tasks.build}
```
Comment (dieunb) [inline app/controllers/supervisor/subjects_controller.rb]: indent

File: app/models/subject.rb [hunk 2]
```diff
@@ -1,4 +1,12 @@
 class Subject < ActiveRecord::Base
-  validates :name, presence: true, length: {maximum: 50}
-  validates :info, presence: true
+  belongs_to :course
+	
+  has_many :activate_course_subjects, class_name: "CourseSubject", foreign_key: "subject_id", dependent: :destroy
```
Comment (dieunb) [inline app/models/subject.rb]: use `'` instead of `"`

File: app/models/subject.rb [hunk 3]
```diff
@@ -1,4 +1,12 @@
...
+  has_many :courses, through: :activate_course_subjects, source: :course
+  has_many :tasks, dependent: :destroy
+
+  accepts_nested_attributes_for :tasks, :reject_if => lambda { |a| a[:name].blank? }, :allow_destroy => true
```
Comment (dieunb) [inline app/models/subject.rb]:
```
reject_if: {|a| a[:name].blank?}, allow_destroy: true
```

File: app/models/subject.rb [hunk 4]
```diff
@@ -1,4 +1,12 @@
...
+  validates :name, presence: true, length: { maximum: 400 }, uniqueness: {case_sensitive: false}
```
Comment (dieunb) [inline app/models/subject.rb]: remove space before `}` and after `{`

File: app/models/task.rb [hunk 5]
```diff
@@ -1,4 +1,4 @@
 class Task < ActiveRecord::Base
-  validates :name, presence: true, length: {maximum: 50}
-  validates :content, presence: true
+  belongs_to :subject
+  validates :name, presence: true, length: { maximum: 400 }, uniqueness: {case_sensitive: false}
```
Comment (dieunb) [inline app/models/task.rb]: remove space before `}` and after `{`

File: app/controllers/supervisor/subjects_controller.rb [hunk 6]
```diff
@@ -46,7 +47,8 @@ def destroy
 
   private
   def subject_params
-    params.require(:subject).permit(:name, :info)
+    params.require(:subject).permit(:name, :info, :total_level, :status,
+    :tasks_attributes => [:id, :name, :content, :_destroy])
```
Comment (dieunb) [inline app/controllers/supervisor/subjects_controller.rb]: indent

File: app/models/subject.rb [hunk 7]
```diff
@@ -1,4 +1,15 @@
 class Subject < ActiveRecord::Base
-  validates :name, presence: true, length: {maximum: 50}
-  validates :info, presence: true
+  belongs_to :course
+	
+  has_many :activate_course_subjects, class_name: 'CourseSubject',
+  foreign_key: '  subject_id', dependent: :destroy
```
Comment (dieunb) [inline app/models/subject.rb]: remove space `'    subject_id'`

File: app/models/subject.rb [hunk 8]
```diff
@@ -1,4 +1,15 @@
...
+  has_many :courses, through: :activate_course_subjects, source: :course
+  has_many :tasks, dependent: :destroy
+
+  accepts_nested_attributes_for :tasks,
+  reject_if: { |a| a[:name].blank? }, :allow_destroy => true
```
Comment (dieunb) [inline app/models/subject.rb]: indent

File: app/models/subject.rb [hunk 9]
```diff
@@ -1,4 +1,15 @@
...
+  validates :name, presence: true, length: {maximum: 400},
+  uniqueness: {case_sensitive: false}
```
Comment (dieunb) [inline app/models/subject.rb]: indent

File: app/views/supervisor/subjects/new.html.erb [hunk 10]
```diff
@@ -2,17 +2,35 @@
...
-      <%= f.text_field :info, class: "form-control" %>
-
-      <%= f.submit "Create subject" , class: "btn btn-primary" %>
-    <% end %>
-  </div>
+  <%= form_for([:supervisor, @subject]) do |f| %>
+    <%= render 'shared/error_messages', object: f.object %>
+    <div class="col-md-6 col-md-offset-0">
+      <div  class="form-group"  >
+    	  <%= f.label :name %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/new.html.erb]: indent all lines code below

File: app/models/subject.rb [hunk 11]
```diff
@@ -1,4 +1,15 @@
...
-  validates :info, presence: true
+  belongs_to :course
+	
+  has_many :activate_course_subjects, class_name: 'CourseSubject',
+    foreign_key: 'subject_id', dependent: :destroy
+  has_many :courses, through: :activate_course_subjects, source: :course
+  has_many :tasks, dependent: :destroy
+
+  accepts_nested_attributes_for :tasks,
+    reject_if: {|a| a[:name].blank?}, :allow_destroy => true
```
Comment (dieunb) [inline app/models/subject.rb]: `allow_destroy: true`

File: app/views/supervisor/subjects/edit.html.erb [hunk 12]
```diff
@@ -2,17 +2,35 @@
...
+    </div>
+    <div class="col-md-6 col-md-offset-0">
+      <%= f.fields_for :tasks do |builder| %>
+        <%= render "task_fields", :f => builder %>
+      <% end %>
+    </div>
+    
+    <p><%= creat_add_task_button f, :tasks %></p>
+    <div class="form-group">
+      <%= f.submit "Edit subject" , class: "btn btn-primary" %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/edit.html.erb]: use as
```
"Edit subject",
```

File: app/views/supervisor/subjects/new.html.erb [hunk 13]
```diff
@@ -2,17 +2,35 @@
...
+    </div>
+    <div class="col-md-6 col-md-offset-0">
+      <%= f.fields_for :tasks do |builder| %>
+        <%= render "task_fields", :f => builder %>
+      <% end %>
+    </div>
+  
+    <p><%= creat_add_task_button f, :tasks %></p>
+    <div class="form-group">
+  	  <%= f.submit "Create subject", class: "btn btn-primary" %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/new.html.erb]: indent

File: app/controllers/supervisor/subjects_controller.rb [hunk 14]
```diff
@@ -46,7 +47,8 @@ def destroy
 
   private
   def subject_params
-    params.require(:subject).permit(:name, :info)
+    params.require(:subject).permit(:name, :info, :total_level, :status,
```
Comment (chuanhtuan) [inline app/controllers/supervisor/subjects_controller.rb]: subject thì không có status
task thì chỉ cần name là đủ

File: app/models/subject.rb [hunk 15]
```diff
@@ -1,4 +1,15 @@
 class Subject < ActiveRecord::Base
-  validates :name, presence: true, length: {maximum: 50}
-  validates :info, presence: true
+  belongs_to :course
+	
+  has_many :activate_course_subjects, class_name: 'CourseSubject',
```
Comment (chuanhtuan) [inline app/models/subject.rb]: remove

File: app/models/subject.rb [hunk 16]
```diff
@@ -1,4 +1,15 @@
...
+    foreign_key: 'subject_id', dependent: :destroy
+  has_many :courses, through: :activate_course_subjects, source: :course
```
Comment (chuanhtuan) [inline app/models/subject.rb]: remove

File: app/views/supervisor/subjects/edit.html.erb [hunk 17]
```diff
@@ -2,17 +2,35 @@
...
-      <%= f.label :name %>
-      <%= f.text_field :name, class: "form-control" %>
-
-      <%= f.label :information %>
-      <%= f.text_field :info, class: "form-control" %>
-      
-      <%= f.submit "Edit Subject", class: "btn btn-primary" %>
-    <% end %>
-  </div>
+  <%= form_for([:supervisor, @subject]) do |f| %>
```
Comment (chuanhtuan) [inline app/views/supervisor/subjects/edit.html.erb]: remove ()

File: app/views/supervisor/subjects/new.html.erb [hunk 18]
```diff
@@ -2,17 +2,34 @@
...
-      <%= f.label :name %>
-      <%= f.text_field :name, class: "form-control" %>
-
-      <%= f.label :info, 'Information' %>
-      <%= f.text_field :info, class: "form-control" %>
-
-      <%= f.submit "Create subject" , class: "btn btn-primary" %>
-    <% end %>
-  </div>
+  <%= form_for([:supervisor, @subject]) do |f| %>
```
Comment (chuanhtuan) [inline app/views/supervisor/subjects/new.html.erb]: remove ()

**Pull request merged**